src/libraries/Common/tests/System/Net/Prerequisites/NetCoreServer/Helpers/RequestInformation.cs

```diff
@@ -46,22 +53,29 @@ public static RequestInformation Create(HttpRequest request)
             }
             info.Cookies = cookies;
 
-            Stream stream = request.GetBufferedInputStream();
-            using (var reader = new StreamReader(stream))
+            string body = string.Empty;
+            try
+            {
+                Stream stream = request.Body;
+                using (var reader = new StreamReader(stream))
+                {
+                    body = await reader.ReadToEndAsync();
+                }
+            }
+            catch (Exception ex)
+            {
+                // We might want to log these exceptions also.
+                body = ex.ToString();
+            }
+            finally
             {
-                string body = reader.ReadToEnd();
                 info.BodyContent = body;
                 info.BodyLength = body.Length;
             }
 
-            info.SecureConnection = request.IsSecureConnection;
+            info.SecureConnection = request.IsHttps;
 
-            var cs = request.ClientCertificate;
-            info.ClientCertificatePresent = cs.IsPresent;
-            if (cs.IsPresent)
-            {
-                info.ClientCertificate = request.ClientCertificate;
-            }
+            //info.ClientCertificate = request.ClientCertificate;
```

Comment:
I wonder how to do this on kestrel. Not doing it makes this [test](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Net.Http.WinHttpHandler/tests/FunctionalTests/ClientCertificateTest.cs#L94) always pass ?

Reply:
Here is [sample code](https://github.com/dotnet/aspnetcore/blob/9ce4a970a21bace3fb262da9591ed52359309592/src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample/Program.cs#L25) for .NET 5
It would require binding to separate port it seems.

Reply:
@karelz could you guys please help me to implement capturing of optional client cert ? .
The [repo](https://github.com/davidsh/corefx-net-endpoint) you mentioned didn't implement it and I don't have the knowledge to do this quickly.
Alternatively we could as well open issue for this and park it, as this doesn't really block my use-case. 

Please let me know, thanks.

Reply:
I don't have a ton of context about this question, but Kestrel allows you to spepcify the new ServerOptionsSelectionCallback with [UseHttps](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.listenoptionshttpsextensions.usehttps?view=aspnetcore-5.0#Microsoft_AspNetCore_Hosting_ListenOptionsHttpsExtensions_UseHttps_Microsoft_AspNetCore_Server_Kestrel_Core_ListenOptions_System_Net_Security_ServerOptionsSelectionCallback_System_Object_).

Reply:
And if you still want to use HttpConnectionAdapterOptions for some reason, you set the [ServerCertificateSelector](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.https.httpsconnectionadapteroptions.servercertificateselector?view=aspnetcore-5.0#Microsoft_AspNetCore_Server_Kestrel_Https_HttpsConnectionAdapterOptions_ServerCertificateSelector) to a custom `Func<ConnectionContext, string, X509Certificate2>`.

Reply:
Thanks @halter73. I spoke with @wfurt yesterday and we agreed that this gap is not blocker for my current xharness/wasm work. So I created https://github.com/dotnet/runtime/issues/52693 to make clear what are the gaps.